pkg/securitypolicy/securitypolicyenforcer.go

```diff
@@ -447,6 +470,77 @@ func possibleIndexesForID(containerID string, mapping map[int]map[string]struct{
 	return possibles
 }
 
+// EnforceExpectedMountsPolicy for StandardSecurityPolicyEnforcer injects a
+// hooks.CreateRuntime hook into container spec and the hook ensures that
+// the expected mounts appear prior container start. At the moment enforcement
+// is expected to take place inside LCOW UVM.
+func (pe *StandardSecurityPolicyEnforcer) EnforceExpectedMountsPolicy(containerID string, spec *oci.Spec) error {
+	pe.mutex.Lock()
+	defer pe.mutex.Unlock()
+
+	if len(pe.Containers) < 1 {
+		return errors.New("policy doesn't allow mounting containers")
+	}
+
+	sandboxID := spec.Annotations[annotations.KubernetesSandboxID]
+	if sandboxID == "" {
+		return errors.New("no sandbox ID present in spec annotations")
+	}
+
+	var wMounts []string
+	pIndices := possibleIndexesForID(containerID, pe.ContainerIndexToContainerIds)
+	if len(pIndices) == 0 {
+		return errors.New("no valid container indices found")
+	}
+
+	matchFound := false
+	for _, index := range pIndices {
+		if !matchFound {
+			matchFound = true
+			wMounts = pe.Containers[index].ExpectedMounts
+		} else {
+			pe.narrowMatchesForContainerIndex(index, containerID)
+		}
+	}
+
+	if len(wMounts) == 0 {
+		return nil
+	}
+
+	var wPaths []string
+	for _, mount := range wMounts {
+		wp := ""
+		if strings.HasPrefix(mount, guestpath.SandboxMountPrefix) {
+			sandboxPath := strings.TrimPrefix(mount, guestpath.SandboxMountPrefix)
+			wp = filepath.Join(guestpath.LCOWRootPrefixInUVM, "sandboxMounts", sandboxPath)
+		} else {
+			// Find the corresponding sandbox mount and resolve the path inside UVM.
```

Comment:
This comment is confusing since we've already handled the sandbox mounts in the if statement above

Reply:
I probably should expand the comment a bit. The hooks run in the runtime space, essentially on the host (UVM). We support two different scenarios here:
1. wait path is provided as a path inside the sandbox mount and should resolve on the host (UVM), e.g., `"sandbox://path/on/the/host"`
2. wait path is provided as a path under a sandbox mount, e.g., `"/sandbox/mount/wait/path/inside/container"`

To handle the first case, we need to substitute the `"sandbox://"` bit with the actual path inside UVM (`"/run/gcs/c/sandboxMounts"`) to get the final path on the host: `"/run/gcs/c/sandboxMounts/path/on/the/host"`

To handle the latter is less straightforward, since we first need to find the corresponding sandbox mount in the OCI spec.
Sandbox mounts are added in the following form:
```
{
    "source": "/run/gcs/c/sandboxMounts/path/on/the/host",
    "destination": "/sandbox/mount
}
```
Now to figure out where the wait path belongs, we have to compare its prefix against all of the container mounts destinations. Once the match is found we can pre-pend the path on the host, the final wait path becomes `"/run/gcs/c/sandboxMounts/path/on/the/host/wait/path/inside/container"`.

Reply:
How do we know the difference between the latter case and a path at a random location in the container with nothing to do with sandboxMounts? Maybe I'm misunderstanding.

Reply:
Maybe an example would be nice here? Are you talking about a use case when there're following mounts:
```
[{
    "source": "/random/not/sandbox/mount/path",
    "destination": "/sandbox/mount/subpath"
},
{
    "source": "/run/gcs/c/sandboxMounts/path/on/the/host",
    "destination": "/sandbox/mount"
}]
```
and the wait path is something like: "/sandbox/mount/subpath/wait/path"?

If that's the case, I'd consider this caller's misconfiguration?
We could also add an extra check to make sure that "source" part of the mount has a sandboxMount prefix to be sure?

Reply:
I'm thinking more of something like 
```
[{ 
    "source": "/random/not/sandbox/mount/path/in/UVM", 
    "destination": "/random/path/in/container" 
}
```

Reply:
chatted offline, I'm adding another use-case where the wait path can be an arbitrary path inside container. In that case we need to resolve the container rootfs inside UVM and pre-pend it to the wait path.